doc/source/serialization.rst

```diff
@@ -64,6 +47,107 @@ Serialization notes
 
 - Lock objects are mostly unserializable, because copying a lock is meaningless and could cause serious concurrency problems. You may have to come up with a workaround if your object contains a lock.
 
+Custom Serialization
+____________________
+
+Sometimes you may want to customize your serialization process because
+the default serializer used by Ray (pickle5 + cloudpickle) does
+not work for you (fail to serialize some objects, too slow for certain objects, etc.).
+
+There are at least 3 ways to define your custom serialization process:
+
+1. If you want to customize the serialization of a type of objects,
+   and you have access to the code, you can define ``__reduce__``
+   function inside the corresponding class. This is commonly done
+   by most Python libraries. Example code:
+
+.. code-block:: python
+
+  import ray
+  import sqlite3
+
+  ray.init()
+
+  class DBConnection:
+      def __init__(self, path):
+          self.path = path
+          self.conn = sqlite3.connect(path)
+
+      # without '__reduce__', the instance is unserializable.
+      def __reduce__(self):
+          deserializer = DBConnection
+          serialized_data = (self.path,)
+          return deserializer, serialized_data
+
+  original = DBConnection("/tmp/db")
+  print(original.conn)
+
+  copied = ray.get(ray.put(original))
+  print(copied.conn)
+
+2. If you want to customize the serialization of a type of objects,
+   but you cannot access or modify the corresponding class, you can
+   register the class with the reducer/serializer you use:
```

Comment:
What are the guarantees if I call `ray.serialization.register_serializer(A, custom_serializer, custom_deserializer)`?

In particular
1. Will the serializer/deserializer be applied on objects of type `A` that are serialized/deserialized on any worker?
2. What about subclasses of `A`?
3. Do we guarantee no race conditions if I call `register_serializer` and then immediately pass an object of type `A` into a remote function? What if it's passed in from a different process?

Also
- If we have this API, we should force keyword arguments (especially if we're doing something like allowing a serializer/deserializer or reducer)

Reply:
Good point, I should include more details in the documentation.

1. It only works on the worker that registers the serializer. Users just need to run the same code path to register the serializer. This is much more robust to worker failure/restarting and avoids concurrency issues caused by importing threads, and no broadcasting is required like before.
2. For subclasses, the behavior is exactly the same as custom serializer in pickle (and same as the previous version): https://python.readthedocs.io/en/latest/library/pickle.html#pickle-dispatch
3. There are no race conditions, because the information of the serializer is provided by the caller who registers the serializer.
4. It could be a good idea to enforce keyword arguments.